src/backend/distributed/commands/type.c

```diff
@@ -0,0 +1,852 @@
+/*-------------------------------------------------------------------------
+ *
+ * type.c
+ *    Commands for TYPE statements.
+ *    The following types are supported in citus
+ *     - Composite Types
+ *     - Enum Types
+ *     - Array Types
+ *
+ *    Base types are more complex and often involve c code from extensions.
+ *    These types should be created by creating the extension on all the
+ *    workers as well. Therefore types created during the creation of an
+ *    extension are not propagated to the worker nodes.
+ *
+ *    Types will be created on the workers during the following situations:
+ *     - on type creation (except if called in a transaction)
+ *       By not distributing types directly when in a transaction allows
+ *       the type to be used in a newly created table that will be
+ *       distributed in the same transaction. In that case the type will be
+ *       created just-in-time to allow citus' parallelism to work.
+ *     - just-in-time
+ *       When the type is not already distributed but used in an object
+ *       that will distribute now. This allows distributed tables to use
+ *       types that have not yet been propagated, either due to the
+ *       transaction case abvove, or due to a type predating the citus
+ *       extension.
+ *     - node activation
+ *       Together with all objects that are marked as distributed in citus
+ *       types will be created during the activation of a new node to allow
+ *       reference tables to use this type.
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/htup_details.h"
+#include "access/xact.h"
+#include "catalog/namespace.h"
+#include "catalog/pg_enum.h"
+#include "catalog/pg_type.h"
+#include "miscadmin.h"
+#include "nodes/makefuncs.h"
+#include "parser/parse_type.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+#include "utils/regproc.h"
+#include "utils/syscache.h"
+#include "utils/typcache.h"
+
+#include "distributed/dist_catalog/distobject.h"
```

Comment:
Again minor: We typically inline `distributed/` headers to above with alphabetical order